### ISSUE TITLE
TLS certificates not associated with Ingress

### ISSUE THREAD

**Issue by eengelking** [coder/enterprise-helm#150]
Today, I spun up a new instance using Helm chart version `1.22`.  

Here are the values.yaml I used for deployment:

```
coderd:
  replicas: 3
  serviceSpec:
    type: ClusterIP
    externalTrafficPolicy: null
    loadBalancerIP: null
    loadBalancerSourceRanges: null
  trustProxyIP: true  
  devurlsHost: "*.<domainName>"
  tls:
    devurlsHostSecretName: coder-certs
    hostSecretName: coder-certs

postgres:
  host: "<databaseDNS>"
  port: 5432
  user: "<databaseUser>"
  database: "<databaseName>"
  passwordSecret: "coder-pg-password"
  sslMode: "require"

  default:
    enable: false

logging:
  json: /dev/stderr
  ```
  
  Upon deployment, an Ingress was generated in the Coder namespace with the following configuration:
  
  ```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: coderd-ingress
  namespace: coder
  labels:
    app.kubernetes.io/managed-by: Helm
  annotations:
    meta.helm.sh/release-name: coder
    meta.helm.sh/release-namespace: coder
    nginx.ingress.kubernetes.io/proxy-body-size: '0'
status:
  loadBalancer:
    ingress:
      - hostname: <snip>.us-east-2.elb.amazonaws.com
spec:
  rules:
    - http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: coderd
                port:
                  number: 80
    - host: '*.<snip>'
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: coderd
                port:
                  number: 80
```

After reviewing the codebase, I discovered there is a section in the `ingress.tpl` template file which is supposed to assign the expected TLS variables, but it doesn't look like it's being called correctly in this case:

```
{{- define "coder.ingress.tls" }}
{{- if (merge .Values dict | dig "ingress" "tls" "enable" false) }}
  tls:
    {{- if and .Values.ingress.host .Values.ingress.tls.hostSecretName }}
    - hosts:
      - {{ .Values.ingress.host | quote }}
      secretName: {{ .Values.ingress.tls.hostSecretName }}
    {{- end }}
    {{- if .Values.devurls }}
    {{- if and .Values.devurls.host .Values.ingress.tls.devurlsHostSecretName }}
    - hosts:
      - {{ include "movedValue" (dict "Values" .Values "Key" "coderd.devurlsHost") }}
      secretName: {{ .Values.ingress.tls.devurlsHostSecretName }}
    {{- end }}
    {{- end }}
{{- end }}
{{- end }}
```

Additionally, there is a slight problem I wished bring up.  The way the Ingress deploys, it sets a wildcard for the host unless otherwise specified.  It would be beneficial if this behavior was changed to allow the values.yaml to be populated with a defined domain name and use this to populate the `devurl`, or allow both to be populated seperately.

To get around these issues, I created a separate ingress:

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: coderd-ingress-new
  namespace: coder
  labels:
  annotations:
    nginx.ingress.kubernetes.io/proxy-body-size: '0'
status:
  loadBalancer:
    ingress:
      - hostname: <snip>.us-east-2.elb.amazonaws.com
spec:
  tls:
    - hosts:
        - <domainName>
        - '*.<domainName>'
      secretName: coder-certs
  rules:
    - host: <domainName>
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: coderd
                port:
                  number: 80
    - host: '*.<domainName>'
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: coderd
                port:
                  number: 80
```

**Comment by mterhar** [comment]
Thanks for catching this. The ingress TLS thing is definitely a bug. 

The host name for Dev URLs is often a separate domain name entirely for the purpose of old browser security. We don't assume you'll want the same domain and a subdomain for Dev URLs. 

We also require satellites to be subdomains of the main coder deployment so using an entirely separate domain is required if you use satellites.

**Comment by eengelking** [comment]
@mterhar Thank you for the follow-up.  

To clarify regarding the Dev URLs, I re-read the documentation this morning and didn't find anything which stated the Dev URL should be a separate domain.  We simply assumed the following would be acceptable based on the documentation:

```
Primary Coder URL: https://coder.foo.com
Dev URL wildcard: https://*.coder.foo.com
```

If I read your comment correctly, which I very well may have misinterpreted, it seems to suggest a totally seperate domain should be used.  Example:

```
Primary Coder URL: https://coder.foo.com
Dev URL wildcard: https://*.coder.bar.com
```

Can you clarify which is correct?  Additionally, is there documentation on the Coder.com site which outlines this suggestion in greater detail?  I cannot find anything specific.

The closest thing I could find was the satellite page, which seemed to indicate my initial understanding to be correct, but I wanted to verify.

```
coderd:
  satellite:
    enable: "true"
    primaryURL: "https://coder.example.com"
    accessURL: "https://coder-eu-west.example.com"
  devurlsHost: "*.coder-eu-west.example.com"
```

**Comment by mterhar** [comment]
> @mterhar Thank you for the follow-up.
> [. . .]
> If I read your comment correctly, which I very well may have misinterpreted, it seems to suggest a totally seperate domain should be used. Example:
> 
> ```
> Primary Coder URL: https://coder.foo.com
> Dev URL wildcard: https://*.coder.bar.com
> ```
> 
> Can you clarify which is correct? Additionally, is there documentation on the Coder.com site which outlines this suggestion in greater detail? I cannot find anything specific.

Sorry for the late reply. There are older browsers that leak cookies and assuming browsers are updated to latest isn't something I'd be comfortable assuming. **Latest** in this context is probably a few years old at this point so it may be pointless. 

The attack is something like a coder.foo.com user opens a devurl to a subdomain devurl like test-mike.coder.foo.com and the session token gets sent to the devurl as a cookie. The cookies in coder are scoped to the subdomain that coder lives on so it shouldn't be an issue if browsers are following the current approach for implementing cookies. Sadly the cookie spec isn't clear and for a long time, firefox, IE, and others had different handling for when a subdomain started with a dot or not.

**Comment by ericpaulsen** [comment]
Closed by #227.
